src/main/java/com/mongodb/spark/sql/connector/read/partitioner/ShardedPartitioner.java

```diff
@@ -0,0 +1,223 @@
+ */
+@ApiStatus.Internal
+public final class ShardedPartitioner implements Partitioner {
+
+  public static final String SHARD_KEY_FIELD_LIST_CONFIG = "shard.key.field.list";
+  public static final List<String> SHARD_KEY_FIELD_LIST_DEFAULT = singletonList("_id");
+
+  private static final String CONFIG_DATABASE = "config";
+  private static final String CONFIG_COLLECTIONS = "collections";
+  private static final String CONFIG_CHUNKS = "chunks";
+  private static final String CONFIG_SHARDS = "shards";
+  private static final String NAMESPACE_FIELD = "ns";
+  private static final String UUID_FIELD = "uuid";
+  private static final String ID_FIELD = "_id";
+  private static final String HOST_FIELD = "host";
+  private static final Bson CHUNKS_PROJECTIONS = Projections.include("min", "max", "shard");
+  private static final Bson SHARDS_PROJECTIONS = Projections.include(ID_FIELD, HOST_FIELD);
+  private static final Bson SORTS = Sorts.ascending("min");
+  private static final BsonValue BSON_MIN = new BsonMinKey();
+  private static final BsonValue BSON_MAX = new BsonMaxKey();
+
+  /** Construct an instance */
+  public ShardedPartitioner() {}
+
+  @Override
+  public List<MongoInputPartition> generatePartitions(final ReadConfig readConfig) {
+    List<String> shardKeys = getShardKeys(readConfig);
+
+    LOGGER.info(
+        "Getting shard chunk bounds for '{}' with shard key: {}",
+        readConfig.getNamespace().getFullName(),
+        shardKeys);
+
+    BsonDocument configCollectionMetadata =
+        readConfig.withClient(
+            client ->
+                client
+                    .getDatabase(CONFIG_DATABASE)
+                    .getCollection(CONFIG_COLLECTIONS, BsonDocument.class)
+                    .find(Filters.eq(ID_FIELD, readConfig.getNamespace().getFullName()))
+                    .projection(Projections.include("_id", "timestamp", "uuid", "dropped"))
+                    .first());
+
+    if (configCollectionMetadata == null) {
+      LOGGER.warn(
+          "Collection '{}' does not appear to be sharded, continuing with a single partition. "
+              + "To split the collections into multiple partitions please use a suitable partitioner.",
+          readConfig.getNamespace().getFullName());
+      return new SinglePartitionPartitioner().generatePartitions(readConfig);
+    }
+
+    if (configCollectionMetadata.getBoolean("dropped", BsonBoolean.FALSE).getValue()) {
+      LOGGER.warn(
+          "Collection '{}' has been dropped continuing with a single partition.",
+          readConfig.getNamespace().getFullName());
+      return new SinglePartitionPartitioner().generatePartitions(readConfig);
+    }
+
+    BsonDocument chunksMatchPredicate =
+        configCollectionMetadata.containsKey("timestamp")
+            ? new BsonDocument(UUID_FIELD, configCollectionMetadata.get(UUID_FIELD))
+            : new BsonDocument(NAMESPACE_FIELD, configCollectionMetadata.get(ID_FIELD));
```

Comment:
> * Is it true that config.chunks.uuid actually contains the UUID of the collection to which the chunk belongs, and not the UUID of the chunk itself? This is unexpected to me, but it seems to be the expectation the predicate relies on.

Thats correct

> * Why do we check whether configCollectionMetadata has the timestamp field before using its uuid instead of directly checking whether it has uuid?

That probably can be simplified - will check.

Reply:
Have added a comment - essentially I was using timestamp as that was the differentiator as to when the uuid field was used instead of the namespace field. But I simplified the filter to use an or lookup.